Features

Istanful opened the pull request:
This is a pull request for the entire master branch. Waow

Istanful committed:
Add savon as a dependency

Savon will be used to execute the requests.

Istanful committed:
Add client class

Istanful committed:
Add authenticate request class

The authenticate request will be used to initiate the
authentication flow.

Istanful committed:
Add collect request class

This class will be used for polling the API whilst waiting
for user confirmation.

Istanful committed:
Implement a response class

All responses will inherit from this class.
Since I do not yet know exactly what the responses will look
like this class will provide an easy way to configure that later.

Istanful committed:
Reconfigure client class

The client class was not correctly configured.
We hade to manually change the message tag format as well
as the namespace path.

Istanful committed:
Add possibility to poll authentication

When authenticating bankid a poll loop is needed.
This commit adds a method for automatically polling the endpoint.
The method will yield the provided block with the current polling
status, making it possible to take action accordingly.

Istanful committed:
Add wsdl configuration possibility

Istanful committed:
Rethink polling solution

This commit rethinks the polling solution.
It is now two separate method calls. The reason
for this change is that you may want to do something with
the authenticate response first.

Istanful committed:
Add autostart token to authenticate response

The autostart token will be used to start
the bankid program on the device using the client.

Istanful committed:
Add proper documentation

hovsater commented on README.md:
```diff
@@ -1,8 +1,8 @@
 # CGIParty
+CGIParty is a gem made for integrating against the CGI Group GRP API.
+As of now you can only perform bankid authorisation.
```
Nitpick: I'd write `As of now you can only perform BankID authorization.`.

Istanful replied:
Check

hovsater commented on README.md:
```diff
@@ -1,8 +1,8 @@
-Welcome to your new gem! In this directory, you'll find the files you need to be able to package up your Ruby library into a gem. Put your Ruby code in the file `lib/cgi_party`. To experiment with that code, run `bin/console` for an interactive prompt.
-
-TODO: Delete this and the text above, and describe your gem
+- *You will need a agreement with CGI group in order to user their API. We do not provide this.*
```
Nitpick: `an agreement`.

Istanful replied:
Check

hovsater commented on README.md:
```diff
@@ -20,27 +20,91 @@ Or install it yourself as:
+  # (Optional) The number of seconds a poll operation will be active before timeout. Recommended is 180 seconds.
+  config.collect_polling_timeout = 180
+
+  # (Optional) The number of seconds a poll operation will wait between each call. Recommended 3 seconds.
```
Nitpick: `Recommended is 3 seconds.`

Istanful replied:
Check

hovsater commented on README.md:
```diff
@@ -20,27 +20,91 @@ Or install it yourself as:
-
-client.collect #<= Should only be called once every three seconds
-#=> CGIParty::CollectResponse
+client = CGIParty.client
```
If the client is a singleton does it make sense to store it in a variable? We could just call `CGIParty.client`.

Istanful replied:
It is actually not a singleton. This was merely a wrapper. I've changed it so that you use the class directly instead. :)

hovsater commented on lib/cgi_party/authenticate_request.rb:
```diff
@@ -0,0 +1,29 @@
+
+    def message_hash
+      {
+        display_name: CGIParty.config.display_name,
```
Do you want to couple your logic directly to the configuration object here or would you rather want to pass it in, using dependency injection? Both are valid approaches, but come with different pros/cons.

Istanful replied:
I agree. It's probably better to use dependency injection. I changed the options to an optional keyword argument which will default to the options provided in the configuration file.

Istanful commented on lib/cgi_party/request.rb:
```diff
@@ -0,0 +1,24 @@
+module CGIParty
+  module Request
```
Create an initialize method which always takes `(savon_client, options, *args)`

hovsater commented on lib/cgi_party/response.rb:
```diff
@@ -0,0 +1,24 @@
+    end
+
+    def respond_to_missing?(method_name, include_private = false)
+      fetch_value(method_name) || super
```
To make this more explicit you could do something like `@source_data.values.first.key?(key)`.

hovsater commented on spec/authenticate_request_spec.rb:
```diff
@@ -0,0 +1,47 @@
+
+RSpec.describe CGIParty::AuthenticateRequest do
+  describe ".initialize" do
+    let(:ssn) { "5907269129" }
```
This is just making the test harder to follow. You have test case, why are we trying to dry-out the test and destroying readability as we're doing it?

hovsater commented on spec/authenticate_request_spec.rb:
```diff
@@ -0,0 +1,47 @@
+  describe "#execute" do
+    include Savon::SpecHelper
+
+    before(:all) { savon.mock!   }
```
Pro-tip: You could use `around` to do the `savon.mock!`, `savon.unmock!` logic in one place. See https://relishapp.com/rspec/rspec-core/v/3-7/docs/hooks/around-hooks

hovsater commented on spec/authenticate_request_spec.rb:
```diff
@@ -0,0 +1,47 @@
+    subject { CGIParty::AuthenticateRequest.new(CGIParty::Client.new.savon_client, ssn) }
+    let(:response_xml) { File.read("spec/fixtures/authenticate_response.xml") }
+
+    it "must call the appropiate soap action" do
```
Again, this test is not reading nicely. I'd recommend taking a look at https://robots.thoughtbot.com/four-phase-test and http://xunitpatterns.com/Four%20Phase%20Test.html

hovsater replied:
Nitpick: Instead of using must, we can be definitive. `it calls the appropriate soap action`.

hovsater commented on spec/authenticate_request_spec.rb:
```diff
@@ -0,0 +1,47 @@
+  end
+
+  describe "#execute" do
+    include Savon::SpecHelper
```
Do we need the spec helper here? Are we just expecting messages on the client or are we depending on logic from the Savon gem also? If we're just expecting messages, I'd recommend a testing spy instead. Using [spies](https://relishapp.com/rspec/rspec-mocks/docs/basics/spies) we could also rewrite the test to read more naturally as well.

```ruby
request.execute
expect(savon_client).to have_received(:message).with(:args)
```

Nice, huh?

Istanful replied:
It turned out I need it for a few specs, however most of them could be mocked just like you said. :)

hovsater commented on spec/authenticate_response_spec.rb:
```diff
@@ -0,0 +1,44 @@
+
+    response = CGIParty::AuthenticateResponse.new(response_body)
+
+    expect(response.transaction_id).to eq response_body[:authenticate_response][:transaction_id]
```
I think you can use the RSpec `have_attributes` matcher here. See https://relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/have-attributes-matcher

hovsater commented on spec/authenticate_response_spec.rb:
```diff
@@ -0,0 +1,44 @@
+      expect(subject).to include authenticate_response.auto_start_token
+    end
+
+    it "must use the provided return url in the query parameterss" do
```
Nitpick: `parameters`

hovsater commented on spec/cgi_party_spec.rb:
```diff
@@ -1,5 +1,9 @@
     expect(CGIParty::VERSION).not_to be nil
   end
+
+  it "must have a wsdl path" do
```
Nitpick: `WSDL`

hovsater commented on spec/client_spec.rb:
```diff
@@ -0,0 +1,89 @@
+RSpec.describe CGIParty::Client do
+  def collect_response_with_status(status)
```
I usually put my helpers at the bottom. It really doesn't matter. Just a personal preference I guess. 🙂

hovsater commented on spec/client_spec.rb:
```diff
@@ -0,0 +1,89 @@
+
+    it "must timeout" do
+      expect(client).to receive(:collect).ordered.once # Initial collect
+      expect(client).to receive(:polling_duration).and_return(CGIParty.config.collect_polling_timeout)
```
I think it will read more clearly if we use something like [timecop](https://github.com/travisjeffery/timecop) to imitate moving forward in time.

hovsater commented on spec/client_spec.rb:
```diff
@@ -0,0 +1,89 @@
+    end
+
+    it "must call collect until authentication is considered finished" do
+      expect(client).to receive(:collect).exactly(3).times
```
I'd much rather use `allow(client).to receive(:collect).and_return()` and then use `expect(client).to have_received(:collect).exactly(3).times`. It reads more naturally.

hovsater commented on spec/client_spec.rb:
```diff
@@ -0,0 +1,89 @@
+          collect_response_with_status("OUTSTANDING_TRANSACTION"),
+          collect_response_with_status("COMPLETE")
+        )
+      expect(client).to receive(:sleep).ordered.with(CGIParty.config.collect_polling_delay)
```
Timecop to the resque! 🎉

hovsater commented on spec/client_spec.rb:
```diff
@@ -0,0 +1,89 @@
+    let(:client) { CGIParty::Client.new }
+
+    it "must do an authenticate request" do
+      expect_any_instance_of(CGIParty::AuthenticateRequest)
```
`expect_any_instance_of` should be considered a smell. There are probably better ways to write this. If you want suggestions, hit me up and I'll give you some suggestions.

hovsater commented on spec/client_spec.rb:
```diff
@@ -0,0 +1,89 @@
+    end
+  end
+
+  describe "#authenticate" do
```
You probably don't even have to test this since they're merely helpers.

hovsater commented on spec/client_spec.rb:
```diff
@@ -0,0 +1,89 @@
+    end
+  end
+
+  describe "#collect" do
```
You probably don't even have to test this since they're merely helpers.

hovsater reviewed (approved):
Awesome work. A few things to work out before this is good to go, but overall it was really well implemented. 👍  🙂

Istanful committed:
fixup! Add proper documentation

Istanful committed:
Add possibility to provide options for a request

You can now explicitly provide options for a request.
If no options are provided it will default to the options
in the fconfiguration.

Istanful committed:
Refactor specs

Istanful committed:
fixup! Implement a response class

Istanful committed:
fixup! Refactor specs

Istanful commented:
@KevinSjoberg I've corrected according to your comments. Merging since barely no functionality have changed, and I know you have other stuff to do. :)
